Fixes #634 - Implemented data entry date option for TS data retrieval

zack-rma opened the pull request:
Fixes #634 - Implements data entry date as option for TimeSeries data retrieval - Serialization bug in progress

zack-rma committed:
Implemented data entry data option for TS data retrieval, serialization bug in progress

MikeNeilson commented on cwms-data-api/src/main/java/cwms/cda/api/TimeSeriesController.java:
```diff
@@ -84,7 +83,7 @@
 public class TimeSeriesController implements CrudHandler {
     private static final Logger logger = Logger.getLogger(TimeSeriesController.class.getName());
-
+    private static final String INCLUDE_ENTRY_DATE = "include-entry-date";
```
From a purely pedantic standpoint this should really be a parameter of the content-type, but I may have to accept the reality of this being easier for everyone. 

@krowvin, we were just discussing this conceptually.

adamkorynta replied:
Path parameters are the "what" I am retrieving, query parameters are the "how" am I adding to or filtering that data, and content-type is the "shape" that is returned. This change effects the how and given the way the column names and data array are paired together to already give this flexibility, I don't see a change in shape.

MikeNeilson replied:
Adding a column is definitely a change in shape. The what is a time series, the query parameters specify exactly which time series, or at least which portion of a given time series (I guess if we're being *really* pedantic begin and end should be in the fragment... but I digress). 

While the flexibility *is* there, it's flexibility to change the shape. I don't totally disagree with you but given we haven't communicated that portion of the contract very well we are introducing a breaking change. We already have more than one downstream library dependent on these types.

I'm going to type up something on the wiki, or maybe discussion, for the philosophy I'm going for with these, hopefully my argument makes more sense in regards to query vs content type. Especially as it relates to some of the challenges we're currently seeing.

adamkorynta replied:
Yeah, it is not conveyed very well that one should use the columns array to determine which index to grab from the data array. I don't think that adding information to the swagger docs to communicate that should trigger a content-type change though (maybe version=2.5 but that seems like a huge headache) especially since not including the parameter to retrieve entry date keeps the array intact for backwards compatibility.

Also, I've never seen an API where the content-type changes how much extra (or less) data gets returned to the client. I'd like to see some examples. I also don't see the reason (pedantic or not) for adding begin/end as path parameters as those are filters on the time series. Everything for the identifier of the time series encompasses the time series (which is also why the date version shouldn't be in the path and is a query parameter).

MikeNeilson replied:
NOTE: not arguing for it, but explanation of my logic on the fragement.
a time series is:
`/timeseries/Alder Springs.Precip-INC.Total.1Hour.1Hour.Calc?version_date=unversioned` identifies a specific time series which is a mathematical series of data - and technically all of it. The entire series could be consider a "document", a fragment is a section within a document; traditionally we would think of lines in a file, but it applies to the time series as well. A file, after all, is just a series of lines.

As I said, extremely pedantic.... admittedly almost to the point of being useless because literally no one does it that way, nor would they even if it could be proven objectively correct.

Technically the units are also representation and not identification but given how limited the use of content-type features are used it would be incredibly difficult to get people to use it; I don't even think the Swagger-UI has a mechanism to slightly tweak the content-type.

But back on the topic of what's correct for us, it seems we're all in agreement that @DanielTOsborne 's initial design is already sufficiently flexible in the current scheme and our failure was in how we documented that for the general end user.

So we leave the inclusion as a query parameter unless a better way actually comes along.

MikeNeilson commented on cwms-data-api/src/main/java/cwms/cda/data/dto/TimeSeries.java:
```diff
@@ -248,6 +256,11 @@ public static class Record {
         @JsonProperty(value = "quality-code", index = 2)
         int qualityCode;
 
```
This might be better as a sub class. I know that adds some complex but eventually we may also want to include the version date and text notes that may be attached and that would be a lot of logic in this one class.

zack-rma replied:
Changed to a subclass. I added a custom deserializer to handle the different classes.

MikeNeilson replied:
Okay, so I said this, Adam said this, but after reading the code and our other discussions above timeseries does it make more sense to just make TimeSeries more generic?

e.g.

A builder where you manually add column names, index, and type, and and functions in the row builder to set such?
something like

```
withColumn(int index, String name, String description, Class<T> type) {
"logic"
}

... Record:

<T> setColumn(int index, T value, Class<T> type) {
  "logic"
}
```
Or something like that, it would prevent the need in TimeSeriesDaoImpl have two different loops doing almost 90% the same work. just a check for "I have this column requested, let's also add it."

Basically instead of hard coding the columns at all (okay, maybe time... is a time series) , the user of the given TimeSeries object (after set by builder) can define them at run-time.


Sorry, I know you did a lot here, that just came to me now looking through the current PR.

MikeNeilson replied:
I don't want Zach to do it in this PR, already been too long, but does anyone think the above idea is crazy?

zack-rma committed:
Created subclass of TimeSeries Record with custom deserializer to handle custom output, adds data entry date support

zack-rma committed:
Fixes build error

MikeNeilson commented:
@jbkolze what are your thoughts on this? At least how it's done. Some appears it may be a breaking change which we want to avoid.

MikeNeilson commented on cwms-data-api/src/main/java/cwms/cda/data/dao/LocationLevelsDaoImpl.java:
```diff
@@ -623,7 +622,7 @@ private static TimeSeries buildTimeSeries(ILocationLevelRef levelRef, Interval i
                 quality = qualityCode.intValue();
             }
-            timeSeries.addValue(dateTime, value, quality);
+            timeSeries.addValue(dateTime, value, quality, null);
```
Are we sure this isn't a breaking change?

zack-rma replied:
I'll double check with additional test cases, but this change should not break any existing functionality. A null data entry date parameter is treated as if a standard Time-Value-Quality data entry was provided. The implementation of `addValue` will use a TimeSeries data record with only three input fields under normal circumstances: 

` if (dataEntryDate != null) {
                values.add(new TimeSeriesRecordWithDate(dateTime, value, qualityCode, dataEntryDate));
            } else {
                values.add(new Record(dateTime, value, qualityCode));
            }`

The existing use cases of TimeSeries should be unaffected, as they will be handled exactly as they were before these changes.

adamkorynta replied:
I recommend subclassing TimeSeries instead

MikeNeilson replied:
I think it would break CWMS.js, The javascript openapi generator already has trouble with our TimeSeries class given some specific assumptions the generator chose to make.

jbkolze replied:
> I think it would break CWMS.js, The javascript openapi generator already has trouble with our TimeSeries class given some specific assumptions the generator chose to make.

Just chiming back in here to add some context... We're already modifying the OpenAPI spec in the cwmsjs build process to adjust a few things, so we can just handle this as well if need be.  I don't think it will cause an issue (not a new one, at least), though, because at this point cwmsjs just anticipates an array of numbers in the response and leaves it up to the user to determine what is what based on either the request or the response metadata.

jbkolze commented:
> @jbkolze what are your thoughts on this? At least how it's done. Some appears it may be a breaking change which we want to avoid.

Conceptually, I don't personally have any qualms with it.  You had mentioned in the typing discussion that you all were trying to leave flexibility for dynamically adjusting the time series values array, and this seems like a prime use case for that.

That being said, I am a little concerned about the part you marked as a breaking change.  I don't know the CDA source that well, but is this indicating that the response would include a fourth null value even if the include_entry_date is false?  Because that definitely would not be ideal -- we've written a lot of CDA code already (and I think other districts have as well) that would have to be updated.  Not as big of a deal if an API version were included in the path (`.../v3/...`), but somewhat cumbersome in the current setup.

My understanding from previous conversations is that you'd get the normal 3-value array if this were set to false, but receive a 4-value array if include_entry_date is true.  And the "value-columns" object would be updated to match.  That'd be my "ideal" implementation.

zack-rma commented:
You are correct that setting the include_entry_date parameter to true would result in a four-value array, whereas setting it to false would return a three-value array. Your "ideal" implementation is what I was aiming for to retain backwards compatibility and avoid breaking any of the other endpoints that rely on the TimeSeries implementation.

zack-rma committed:
Fixes test case failure

adamkorynta commented on cwms-data-api/src/main/java/cwms/cda/data/dao/TimeSeriesDaoImpl.java:
```diff
@@ -391,14 +393,55 @@ public TimeSeries getTimeseries(String page, int pageSize, String names, String
-                            tsRecord.getValue(qualityNormCol).intValue()
-                    )
-            );
+            if (includeEntryDate) {
```
This query is doubling the time it takes to retrieve time series. Can this replace the `retrieve_ts_out_tab` calls?

zack-rma replied:
While it could replace the `retrieve_ts_out_tab` call above, doing so would require implementing trim support into the query, as that is currently handled by the `retrieve_ts_out_tab` call. I haven't quite figured out the best way to do so, so maybe we can discuss this in more detail

adamkorynta replied:
if the schema needs to provide more, we can update the schema call

zack-rma replied:
I've created a PR with a new schema call that returns timeseries data with the data entry dates: https://bitbucket.hecdev.net/projects/CWMS/repos/cwms_database/pull-requests/412/overview

adamkorynta commented on cwms-data-api/src/main/java/cwms/cda/data/dto/TimeSeries.java:
```diff
@@ -159,7 +164,8 @@ public ZonedDateTime getEnd() {
     // Use the array shape to optimize data transfer to client
-    @JsonFormat(shape=JsonFormat.Shape.ARRAY)
+    @JsonFormat(shape = JsonFormat.Shape.ARRAY)
+    @JsonDeserialize(contentUsing = TimeSeriesRecordDeserializer.class)
```
This method is overwritten for XML using a Mixin, did you verify that behavior works as intended still?

zack-rma replied:
I added a Mixin test that verifies that the XML tags for the value records have the appropriate labels. There are also a couple serialization/deserialization tests that verify that this works as intended.

adamkorynta commented on cwms-data-api/src/main/java/cwms/cda/data/dto/TimeSeries.java:
```diff
@@ -186,30 +191,35 @@ public ZonedDateTime getVersionDate() {
     @Schema(name = "value-columns", accessMode = AccessMode.READ_ONLY)
     public List<Column> getValueColumnsJSON() {
-        return getColumnDescriptor();
+        return getColumnDescriptor((values != null && !values.isEmpty())
```
This seems like behavior for a subclass

zack-rma replied:
Added to TimeSeries subclass

adamkorynta commented on cwms-data-api/src/main/java/cwms/cda/data/dto/TimeSeries.java:
```diff
@@ -218,7 +228,16 @@ private List<Column> getColumnDescriptor() {
             }
         }
-
+        if (includeDataEntryDate) {
```
This could also be accomplished better with a subclass

zack-rma replied:
Moved into subclass

adamkorynta commented on cwms-data-api/src/main/java/cwms/cda/data/dto/TimeSeries.java:
```diff
@@ -240,6 +259,16 @@ private List<Column> getColumnDescriptor() {
+
+    // This class is used to deserialize the time-series data JSON into an object
+    // Solves the issue of the deserializer getting stuck in a loop
+    // and throwing a StackOverflowError when trying to handle the Record class directly
```
This seems sketchy to me, why is your custom deserializer causing this?

zack-rma replied:
Removed custom deserializer

adamkorynta commented on cwms-data-api/src/main/java/cwms/cda/formatters/json/adapters/TimeSeriesRecordDeserializer.java:
```diff
@@ -0,0 +1,77 @@
+            return jsonParser.getCodec().treeToValue(node, TimeSeriesRecordWithDate.class);
+        }
+        String nodeString = node.toString();
+        if (nodeString.startsWith("[")) {
```
A mixin doesn't solve the need for this custom parsing? All this logic looks like we're circumventing jackson too much

zack-rma replied:
Removed custom serializer

adamkorynta commented on cwms-data-api/src/main/java/cwms/cda/formatters/json/adapters/TimeSeriesRecordDeserializer.java:
```diff
@@ -0,0 +1,77 @@
+                Timestamp dateTime = Timestamp.from(Instant.ofEpochMilli(Long.parseLong(valList[0])));
+                double value = Double.parseDouble(valList[1]);
+                int quality = Integer.parseInt(valList[2]);
+                Timestamp entryDate = Timestamp.from(Instant.ofEpochMilli(Long.parseLong(valList[3])));
```
no need to convert from Instant to Timestamp, Timestamp's constructor takes in epoch millis.

adamkorynta commented on cwms-data-api/src/main/java/cwms/cda/formatters/json/adapters/TimeSeriesRecordDeserializer.java:
```diff
@@ -0,0 +1,77 @@
+    @Override
+    public TimeSeries.Record deserialize(JsonParser jsonParser, DeserializationContext deserializationContext) throws IOException {
+        JsonNode node = jsonParser.readValueAsTree();
+        if (node.get("data-entry-date") != null) {
```
This should be a constant

MikeNeilson replied:
should probably also be ignored, data-entry-date is *always* set by the database itself, an external system/user isn't allowed to change it.

MikeNeilson commented:
> 
> My understanding from previous conversations is that you'd get the normal 3-value array if this were set to false, but receive a 4-value array if include_entry_date is true. And the "value-columns" object would be updated to match. That'd be my "ideal" implementation.

That is correct for the location indicated, which is location levels backed by a time series. I don't know how much it would affect what you're currently using, but it's definitely not ideal. At the least it definitely *shouldn't* be null, may as well provide the data, but seems like a parameter should be added to match on the location level end point.

But it sounds like your on the same page with Adam about the shape being already explicitly flexible so I'm okay with that section now; not "ideal", but what is, definitely something to better document though.

zack-rma committed:
634 TimeSeries Subclass update

zack-rma committed:
634 TimeSeries Subclass update - added Mixin test

MikeNeilson commented on cwms-data-api/src/main/java/cwms/cda/data/dao/TimeSeriesDaoImpl.java:
```diff
@@ -391,59 +435,29 @@ public TimeSeries getTimeseries(String page, int pageSize, String names, String
-                });
+                logger.fine(() -> query2.getSQL(ParamType.INLINED));
+                final TimeSeriesWithDate timeSeries = new TimeSeriesWithDate(timeseries);
+                query2.forEach(tsRecord -> timeSeries.addValue(
```
we don't need to solve this now, but definitely before we add requesting any text notes as well. There has to be a better way to handle this with the builders.

MikeNeilson replied:
Given the previous logic above, does this need a new or just a cast so you can get the desired methods?

MikeNeilson commented on cwms-data-api/src/main/java/cwms/cda/data/dao/TimeSeriesDaoImpl.java:
```diff
@@ -351,27 +350,72 @@ public TimeSeries getTimeseries(String page, int pageSize, String names, String
+        TimeSeries retVal = null;
         if (pageSize != 0) {
+            if (versionDate != null) {
+                whereCond = whereCond.and(AV_TSV_DQU.AV_TSV_DQU.VERSION_DATE.eq(versionDate == null ? null :
```
Does this logic handle max version? or will AV_TSV_DQU always return every version? or only the specifically requested version?

zack-rma replied:
Updated logic to handle max version in line with previous implementation. If the version date is not specified, the max version will be set to true and the most recent version date will be used for retrieval. This does prevent retrieval of multiple versions in one call, so that could be a problem if that is a desired feature.

zack-rma committed:
Updated max version handling, added test case

zack-rma commented:
Build errors appear to be unrelated to these changes

adamkorynta commented on cwms-data-api/src/main/java/cwms/cda/api/TimeSeriesController.java:
```diff
@@ -382,6 +381,8 @@ public void delete(@NotNull Context ctx, @NotNull String timeseries) {
                         + "\n* `wml2` (only if name field is specified)"
                         + "\n* `json` (default)"),
+                @OpenApiParam(name = INCLUDE_ENTRY_DATE, type = Boolean.class, description = "Specifies "
+                    + "whether to include the data entry date in the response.  Default is false."),
```
this should very clearly specify that it increases the data array size to include the date

zack-rma replied:
Added comment reflecting the change in response format

adamkorynta commented on cwms-data-api/src/main/java/cwms/cda/data/dto/TimeSeries.java:
The `value-columns` and `records` properties needs much more description. It needs need to describe how they relate to each other and how the records array can be variable length, so using the column index is key.

zack-rma replied:
Added better descriptions, hopefully that clears up the usage

adamkorynta commented on cwms-data-api/src/main/java/cwms/cda/data/dto/TimeSeriesWithDate.java:
```diff
@@ -0,0 +1,177 @@
+@JsonNaming(PropertyNamingStrategies.KebabCaseStrategy.class)
+@FormattableWith(contentType = Formats.JSONV2, formatter = JsonV2.class, aliases = {Formats.DEFAULT, Formats.JSON})
+@FormattableWith(contentType = Formats.XMLV2, formatter = XMLv2.class, aliases = {Formats.XML})
+public final class TimeSeriesWithDate extends TimeSeries {
```
Should be `TimeSeriesWithEntryDate` or `TimeSeriesWithDataEntryDate`. Time series implies there are dates already

zack-rma replied:
Renamed for clarity

adamkorynta commented on cwms-data-api/src/main/java/cwms/cda/data/dto/TimeSeriesWithDate.java:
```diff
@@ -0,0 +1,177 @@
+
+	// list of TimeSeriesWithDate.Record, uses raw to avoid typing errors
+    @Override
+    public List getValues() {
```
typing was left off

zack-rma replied:
Added typing

adamkorynta commented on cwms-data-api/src/main/java/cwms/cda/data/dto/TimeSeriesWithDate.java:
```diff
@@ -0,0 +1,177 @@
+@FormattableWith(contentType = Formats.XMLV2, formatter = XMLv2.class, aliases = {Formats.XML})
+public final class TimeSeriesWithDate extends TimeSeries {
+
+    private List<TimeSeriesWithDate.Record> values;
```
should be able to reuse the parent's values since there is inheritance

zack-rma replied:
Needed to add the local variables to get the typing correct. Removing it causes loss of type within the List, which drops the data entry date value.

adamkorynta replied:
then you shouldn't inherit from TimeSeries as the class structure then has two sets of data

MikeNeilson replied:
This lends itself to the other suggestion I made with the more generic structure. Though I think that would be too much for this PR.

The Time Series class Daniel created is already generic enough (in concept at least) to handle adding additional columns.

zack-rma replied:
Cleaned up the duplicate value list data from the subclass. Required adding custom deserializer back to properly handle the subtypes of TimeSeries.Record.

adamkorynta commented on cwms-data-api/src/main/java/cwms/cda/data/dto/TimeSeriesWithDate.java:
```diff
@@ -0,0 +1,177 @@
+        values = new ArrayList<>();
+    }
+
+    public TimeSeriesWithDate(TimeSeries timeSeries) {
```
Why so many constructors? Isn't there only one endpoint returning only one time series? Also constructors should go at the top of the class

zack-rma replied:
Removed unnecessary constructors, swapped usages where needed

adamkorynta commented on cwms-data-api/src/main/java/cwms/cda/data/dto/TimeSeriesWithDate.java:
```diff
@@ -0,0 +1,177 @@
+        @JsonProperty(value = "data-entry-date", index = 3)
+        @Schema(implementation = Long.class, description = "Milliseconds since 1970-01-01 (Unix Epoch), always UTC")
+        @JsonInclude(JsonInclude.Include.NON_DEFAULT)
+        Timestamp dataEntryDate;
```
private

zack-rma replied:
Made private

adamkorynta commented on cwms-data-api/src/main/java/cwms/cda/api/TimeSeriesController.java:
Post/Patch should throw an error if the client is sending a time series with a data-entry-date as that field isn't editable

zack-rma replied:
Added a check for data entry date values

adamkorynta commented on cwms-data-api/src/test/resources/cwms/cda/api/lrl/1day_offset_version_date.json:
Where's the example JSON with the data entry date included? Serialization unit tests should test against the file.

zack-rma replied:
Added example JSON files and a test for serialization against them

adamkorynta commented on cwms-data-api/src/test/java/cwms/cda/data/dao/TimeSeriesDaoTest.java:
```diff
@@ -171,7 +169,7 @@ public void testCreateWithData() throws Exception
 			// Also, should end be the last point or the next interval?
 
-			TimeSeries ts = new TimeSeries(null, -1, 0, tsId, officeId, start, end, "m", Duration.ofMinutes(minutes));
+			TimeSeriesWithDate ts = new TimeSeriesWithDate(null, -1, 0, tsId, officeId, start, end, "m", Duration.ofMinutes(minutes));
```
Shouldn't be able to store time series with a data-entry-date, no need for this test to change.

zack-rma replied:
Reverted test, test is disabled due to deprecated connection usage

adamkorynta commented on cwms-data-api/src/test/java/cwms/cda/api/TimeseriesControllerTestIT.java:
```diff
@@ -242,6 +349,91 @@ void test_lrl_1day_malicious_units() throws Exception {
+            .log().ifValidationFails(LogDetail.ALL, true)
+            .accept(Formats.JSONV2)
+            .header("Authorization", user.toHeaderValue())
+            .queryParam(Controllers.OFFICE, officeId)
```
where is the query param to include the data entry date?

zack-rma replied:
I added this request to check that we can get the same data back without the data entry date, the request above gets it with the entry date. I moved the query parameter variable to the Controllers class for better accessibility

MikeNeilson replied:
Might be good to set the parameter explicitly to false. There's always the possibility we change the default behavior.

zack-rma committed:
Updated based on Adam's feedback. Added serializer test on file, cleaned up DTO

zack-rma committed:
Updated based on Adam's feedback. Added error when attempting to store values with data entry date

adamkorynta commented on cwms-data-api/src/main/java/cwms/cda/api/TimeSeriesController.java:
```diff
@@ -200,7 +163,8 @@ public void create(@NotNull Context ctx) {
 
             TimeSeriesDao dao = getTimeSeriesDao(dsl);
-            TimeSeries timeSeries = deserializeTimeSeries(ctx);
+            TimeSeriesWithDataEntryDate timeSeries = deserializeTimeSeries(ctx);
```
This should fail deserialization if the time series doesn't include the data entry date since you are typing to the subclass.

zack-rma replied:
I thought so too, but it deserializes with null values for the data entry dates. This seems to be the most reliable way to check for those values, since using the body content as a String or InputStream to check for them would consume the context and close the data stream. I found that the default behavior when passing in a TS with data entry values was to drop/ignore the values and treat it as a standard TS, so typing to the subclass doesn't appear to cause issues since it is cast to the base class anyway for storage. 

I'm open to suggestions for a better way to approach this.

zack-rma replied:
After more digging and some testing, I've managed to find a solution that does not close the data stream. I've removed the deserialization to the child class

adamkorynta commented on cwms-data-api/src/main/java/cwms/cda/data/dto/TimeSeries.java:
```diff
@@ -79,7 +79,8 @@ public class TimeSeries extends CwmsDTOPaginated {
     @ArraySchema(
             schema = @Schema(
-                    description = "List of retrieved time-series values",
+                    description = "List of retrieved time-series values. Contains [dateTime, value, qualityCode]. "
```
The purpose for the description is to allow the array to be unspecified, the description should not identify what the columns are, rather the value-columns property is what tells the client what the columns are. ex, if we add a boolean "include-quality" in the future, we want to allow the array to be of size 2.

zack-rma replied:
I removed the column data description and added a note that the value-columns describe the value array data structure.

adamkorynta commented on cwms-data-api/src/test/resources/cwms/cda/api/lrl/1day_offset_version_date_roundtrip.json:
this file name should specify that it includes data entry dates

zack-rma replied:
Renamed to better identify the data and it's usage

zack-rma committed:
Renamed json file, fixed documentation

zack-rma committed:
Fixed data entry date request body checking

zack-rma committed:
Cleaned up TS retrieval with data entry dates.
Added integration tests with data.
Fixed unit parameter usage in integration tests

zack-rma commented:
Build will fail until database schema changes are merged and a new release is available. Changes can be found here: https://bitbucket.hecdev.net/projects/CWMS/repos/cwms_database/pull-requests/412/overview

zack-rma committed:
Updated comment

MikeNeilson commented:
Tests are still failing, actions output is still showing schema of "latest-dev"

zack-rma committed:
Added null check

inguyen314 commented:
what is status?

zack-rma commented:
> what is status?

The necessary changes to the database were just merged in, we'll need an updated schema to be available before this feature can be finalized.

zack-rma committed:
Added database compatibility check

MikeNeilson commented on cwms-data-api/src/main/java/cwms/cda/data/dao/TimeSeriesDaoImpl.java:
```diff
@@ -352,25 +346,69 @@ public TimeSeries getTimeseries(String page, int pageSize, String names, String
+            }
         });
 
+        if (includeEntryDate) {
```
why are we doing this check if the above would already have returned a TimeSeriesWithDataEntryDate?

Alternatively why have the login within the lambda for data_entry_date if we're goign to change it here?

(I'm not seeing a difference other than the type after new.)

zack-rma replied:
Remnant from before the above line, perhaps. I've removed it and tested that all works as intended without it.

MikeNeilson commented on cwms-data-api/src/main/java/cwms/cda/data/dao/TimeSeriesDaoImpl.java:
```diff
@@ -1127,15 +1184,15 @@ private void store(Connection connection, String officeId, String tsId, String u
-                }
-                tsvArray.add(new ZTSV_TYPE(value.getDateTime(), dataValue, BigDecimal.valueOf(value.getQualityCode())));
-            }
+			for(TimeSeries.Record value : values)
```
incorrect curlies, but otherwise thank you for cleaning up the loop structures and formatting.

zack-rma replied:
Fixed brackets

MikeNeilson commented on cwms-data-api/src/main/java/cwms/cda/data/dto/TimeSeries.java:
```diff
@@ -253,9 +265,11 @@ public static class Record {
+        }
 
         public Record(Timestamp dateTime, Double value, int qualityCode) {
+            super();
```
... what super?

zack-rma replied:
Leftover from when I had created a shared parent class to resolve some serialization issues. Removed

MikeNeilson commented on cwms-data-api/src/main/java/cwms/cda/data/dto/TimeSeriesRecordWithDate.java:
```diff
@@ -0,0 +1,81 @@
+ * The data entry date is the date that the data was entered into the database.
+ */
+@JsonDeserialize(using = JsonDeserializer.None.class)
+public final class TimeSeriesRecordWithDate extends TimeSeries.Record {
```
This confused me. Should be `TimeSeriesRecordWithEntryDate`

zack-rma replied:
Renamed as suggested

MikeNeilson commented on cwms-data-api/src/test/java/cwms/cda/api/TimeseriesControllerTestIT.java:
```diff
@@ -75,10 +78,9 @@ void test_lrl_timeseries_psuedo_reg1hour() throws Exception {
                 .header("Authorization",user.toHeaderValue())
                 .queryParam("office",officeId)
-                .queryParam("units","cfs")
+                .queryParam("unit","cfs")
```
It would seem we should really start using the constants we've defined. (Not a ding, pretty sure we're all guilty of that.)

zack-rma replied:
Replaced with Controllers constant where applicable

MikeNeilson commented on cwms-data-api/src/test/java/cwms/cda/api/TimeseriesControllerTestIT.java:
```diff
@@ -224,24 +444,148 @@ void test_lrl_1day_malicious_units() throws Exception {
+            .accept(Formats.JSONV2)
+            .header("Authorization", user.toHeaderValue())
+            .queryParam(Controllers.OFFICE, officeId)
+            .queryParam(Controllers.UNIT, "CFS")
```
Yes, this is definitely the superior way to set those parameters.

MikeNeilson commented on cwms-data-api/src/test/java/cwms/cda/data/dao/TimeSeriesDaoTest.java:
thank you for updating the formatting within this file while you were at it.

MikeNeilson reviewed (changes requested):
I have some minor nitpicks on formatting.

zack-rma committed:
Replaced strings with constants, cleaned up DAO

zack-rma committed:
Merge branch 'develop' into bugfix/ts_data_entry_issue_634

zack-rma committed:
Fixed scoping issue

zack-rma committed:
Removed duplicate value list, added custom deserializer to handle subtypes of Record class

zack-rma committed:
Renamed class

zack-rma committed:
Fixed broken ratings test

adamkorynta commented on cwms-data-api/src/main/java/cwms/cda/api/TimeSeriesController.java:
```diff
@@ -573,17 +543,24 @@ public void update(@NotNull Context ctx, @NotNull String id) {
     }
 
-    private TimeSeries deserializeTimeSeries(Context ctx) throws IOException {
+    private TimeSeries deserializeTimeSeries(Context ctx) throws IOException
```
brackets don't follow standards

zack-rma replied:
Fixed brackets

zack-rma committed:
Merge branch 'develop' into bugfix/ts_data_entry_issue_634

zack-rma committed:
Cleaned up subclasses, cleaned up deserializer. Added more unit test coverage of ts record classes

adamkorynta commented on cwms-data-api/src/main/java/cwms/cda/formatters/json/adapters/TimeSeriesRecordDeserializer.java:
```diff
@@ -0,0 +1,91 @@
+ * getting stuck in a loop if the Record class is used directly.
+ * Allows for use of subclass with additional fields {@link TimeSeriesRecordWithEntryDate}.
+ */
+public final class TimeSeriesRecordDeserializer extends JsonDeserializer<TimeSeries.Record> {
```
We don't deserialize the time series with data entry date since the client is no allowed to store those entry dates. We shouldn't need this class. 

Taking this logic though, it seems like we can create a very simple custom _serializer_, remove the subclassing of TimeSeries and TimeSeries.Record, move the entry date to the TimeSeries.Record, and conditionally add to the array if the data entry date is not null. There should be no case where the data entry date was requested, but then it is returned as null in the array. This is not the case if we add text in the future, but I'm sure we can figure out a reasonable solution when that feature is requested.

zack-rma replied:
Removed deserializer class and added custom serializer to handle records. Removed subclass of TimeSeries.Record.

adamkorynta reviewed (changes requested):
I think we're almost there

zack-rma committed:
Removed deserializer, added custom serializer for ts records, added and updated tests

zack-rma committed:
Merge branch 'develop' into bugfix/ts_data_entry_issue_634

zack-rma committed:
Updated serializer design, improved unit tests. XML deserialization still in progress

zack-rma committed:
Fixed custom serializer to handle XML serialization
